rl4co/envs/routing/pdp/env.py

```diff
@@ -209,3 +209,317 @@ def check_solution_validity(self, td, actions):
+        return render(td, actions, ax)
+
+
+class PDPRuinRepairEnv(RL4COEnvBase):
+    """Pickup and Delivery Problem (PDP) environment for performing neural rein-repair search.
+    The environment is made of num_loc + 1 locations (cities):
+        - 1 depot
+        - `num_loc` / 2 pickup locations
+        - `num_loc` / 2 delivery locations
+    
+    The goal is to search for optimal solutions to pickup and delivery problems by performing a ruin-and-repair neighborhood search on a given initial solution.
+    (see MDP described in https://arxiv.org/abs/2204.11399)
+    
+    The condition is that at each step, the visited solutions must be feasible, 
+    maintaining the sequence of visiting the pickup location before its corresponding delivery location.
+
+    Observations:
+        - locations of the depot, pickup, and delivery locations
+        - current solution to be improved
+        - historical decisions
+        - the current step
+
+    Constraints:
+        - the tour starts and ends at the depot
+        - each pickup location must be visited before its corresponding delivery location
+        - the vehicle cannot visit the same location twice
+
+    Finish Condition:
+        - None
+
+    Reward:
+        - the immediate reduced cost over the current best-so-far solution
+        (see MDP described in https://arxiv.org/abs/2204.11399)
+        
+    Args:
+        num_loc: number of locations (cities) in the TSP
+        init_sol_type: the method type used for generating initial solutions (random or greedy)
+        td_params: parameters of the environment
+        seed: seed for the environment
+        device: device to use.  Generally, no need to set as tensors are updated on the fly
+    """
+
+    name = "pdp_ruin_repair"
+
+    def __init__(
+        self,
+        generator: PDPGenerator = None,
+        generator_params: dict = {},
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        if generator is None:
+            generator = PDPGenerator(**generator_params)
+        self.generator = generator
+        self._make_spec(self.generator)
+
+    @staticmethod
+    def _step(td: TensorDict) -> TensorDict:
+        
+        # get state information from td
+        action = td["action"]
+        selected = action[:,0].view(-1,1)
+        first = action[:,1].view(-1,1)
+        second = action[:,2].view(-1,1)
+        solution = td["rec_current"]
+        solution_best = td["rec_best"]
+        locs = td["locs"]
+        cost_bsf = td["cost_bsf"]
+        action_record = td["action_record"]
+        bs, gs = solution.size()
+
+        # perform ruin and repair
+        next_rec = PDPRuinRepairEnv._insert_operator(solution, selected + 1, first, second)
+        new_obj = PDPRuinRepairEnv.get_costs(locs, next_rec)
+
+        # compute reward and update best-so-far solutions
+        now_bsf = torch.min(torch.cat((new_obj[:,None], cost_bsf[:, None]),-1),-1)[0]
```

Comment:
[Enhance] This could be replaced to
```python
now_bsf = torch.where(new_obj<cost_bsf, new_obj, cost_bsf)
```
for efficiency and clearness.

Reply:
LOL yes! This code is copied from previous projects and not refined haha